design/authenticators/authn_sut/authn_sut_solution_design.md

```diff
@@ -0,0 +1,384 @@
+salted and hashed using a cryptographically secure hash function, such as SHA-256.
+When a user presents a SUT to authenticate to Conjur, we will hash the SUT and
+compare it to the hash stored in the database.
+
+##### SUT Format
+
+We cannot use JWTs for the SUT as we need the ability to revoke the SUT after it
+is used, a feature not easily implemented with JWTs. This would make it
+impossible to prevent replay attacks.
+
+##### SUT Expiration and Revocation
+
+The SUT will have a short expiration time, on the order of 20-30 seconds. This is
+enough time to allow the client to use the SUT to authenticate to Conjur, even
+in the case of network latency, but short enough to limit the window of
+opportunity for an attacker to steal and use the SUT.
+
+Additionally, only one SUT will be valid at a time for a given user. If a new
+SUT is generated for a user, the previous SUT will be deleted.
+
+When a SUT is used (or revoked), it will be deleted from the database.
+Expired SUTs can also safely be deleted from the database, though
+this is not strictly necessary since they will not be valid past
+their expiration time.
+When deleting used or expired SUTs, we must not cascade the delete
+since this would delete any related audit logs.
+
+##### Code Verifier and Code Challenge
+
+As an additional layer of security, the client will generate a code verifier,
+which can be a Guid or other cryptographically secure random string. The code
+will then be hashed and sent along with the login request to be associated
+with the generated SUT. This hashed code verifier, known as the code
+challenge, will be stored by Conjur along with the SUT. When the client
+attempts to authenticate with the SUT, they will also send the code verifier
+(in plain text form), and Conjur will hash it and compare it to the stored
+code challenge.
+
+Initially we will support the SHA-256 algorithm for the code challenge. In the
+future we may want to support additional algorithms.
+
+The code challenge and code verifier flow are modeled after OIDC PKCE.
+
+##### Open Questions
+
+- Is it possible to restrict use of the SUT to a specific IP range (like with
+  hostfactory tokens) or other criteria to limit risk if the SUT is leaked?
+- Should we provide a way for clients to register a key pair to use for
+  signing the SUT generation request? This would allow us to verify that the
+  request came from a trusted client.
+- Should we create a job to clean up expired SUTs?
+- Should we create a mechanism to revoke SUTs before they expire (other than
+  manually deleting them from the database)?
+
+##### Example Database Table
+
+| id | role_id | guid_hash | code_challenge | code_challenge_algorithm | expires_at |
+|----|---------|-----------|----------------|--------------------------|------------|
+| 1  | user:john.doe | lkduj...  | 3iod... | sha256 | 2023-09-28 12:08:23 |
+| 2  | user:demo.user | 2k1g2...  | kld80... | sha256 | 2023-09-28 12:09:45 |
```

Comment:
A couple of comments

- [semantic] `role_id` should have an account in it.
- I'd recommend changing `guid_hash` to `token_hash` to avoid leaking the implementation details. (I'd recommend avoiding GUIDs in favor of base32 or base58 encoded strings of a pre-determined length).
- The column `id` needs to be designed with care to ensure customers can't exhaust the available values as the rows increment.  It might make more sense to use `token_hash` as the unique key (which guarantees uniqueness of available single-use tokens at any point in time).